service/history/decision/handler_test.go

```diff
@@ -80,34 +109,186 @@ func (s *DecisionHandlerSuite) TearDownTest() {
+			domainID: _testDomainUUID,
+			mutablestate: &persistence.WorkflowMutableState{
+				ExecutionInfo: &persistence.WorkflowExecutionInfo{},
+			},
+			expectCalls: func(shardContext *shard.MockContext) {
+				shardContext.EXPECT().GetEventsCache().Times(1).Return(events.NewMockCache(s.controller))
+			},
+			expectErr: false,
+		},
+		{
+			name:     "test HandleDecisionTaskScheduled - with completed workflow",
+			domainID: _testDomainUUID,
+			mutablestate: &persistence.WorkflowMutableState{
+				ExecutionInfo: &persistence.WorkflowExecutionInfo{
+					// WorkflowStateCompleted = 2 from persistence WorkflowExecutionInfo.IsRunning()
+					State: 2,
+				},
+			},
+			expectCalls: func(shardContext *shard.MockContext) {
+				shardContext.EXPECT().GetEventsCache().Times(1).Return(events.NewMockCache(s.controller))
+			},
+			expectErr: true,
+		},
+		{
+			name:     "test HandleDecisionTaskScheduled - with failure to get start event",
+			domainID: _testDomainUUID,
+			mutablestate: &persistence.WorkflowMutableState{
+				ExecutionInfo: &persistence.WorkflowExecutionInfo{
+					// execution has no event yet
+					DecisionScheduleID: -23,
+					LastProcessedEvent: -23,
+				},
+			},
+			expectCalls: func(shardContext *shard.MockContext) {
+				eventsCache := events.NewMockCache(s.controller)
+				shardContext.EXPECT().GetEventsCache().Times(1).Return(eventsCache)
+				eventsCache.EXPECT().
+					GetEvent(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+					Times(1).
+					Return(nil, &persistence.TimeoutError{Msg: "failed to get start event: request timeout"})
+				shardContext.EXPECT().GetShardID().Return(_testShardID).Times(1)
+			},
+			expectErr: true,
+		},
+		{
+			name:     "test HandleDecisionTaskScheduled - first decision task scheduled failure",
+			domainID: _testDomainUUID,
+			mutablestate: &persistence.WorkflowMutableState{
+				ExecutionInfo: &persistence.WorkflowExecutionInfo{
+					DecisionScheduleID: -23,
+					LastProcessedEvent: -23,
+				},
+				BufferedEvents: append([]*types.HistoryEvent{}, &types.HistoryEvent{}),
+			},
+			expectCalls: func(shardContext *shard.MockContext) {
+				eventsCache := events.NewMockCache(s.controller)
+				shardContext.EXPECT().GetEventsCache().Times(1).Return(eventsCache)
+				eventsCache.EXPECT().
+					GetEvent(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+					Times(1).
+					Return(&types.HistoryEvent{}, nil)
+				shardContext.EXPECT().GetShardID().Return(_testShardID).Times(1)
+				shardContext.EXPECT().GenerateTransferTaskIDs(gomock.Any()).Times(1).Return([]int64{}, errors.New("some random error to avoid going too deep in call stack unrelated to this unit"))
+			},
+			expectErr:       true,
+			isfirstDecision: true,
+		},
+		{
+			name:     "test HandleDecisionTaskScheduled - first decision task scheduled success",
```

Comment:
you can probably avoid "test HandleDecisionTaskScheduled -" prefix in test case names. `t.Run` puts the parent name in test output. You can double check by intentionally failing one of the tests

Reply:
addressed in follow up PR